WooCommerce/Classes/ViewRelated/BottomSheet/ListSelector/BottomSheetListSelectorViewController.swift

```diff
@@ -0,0 +1,127 @@
+import UIKit
+
+/// Displays a list (implemented by table view) for the user to select a generic model from a bottom sheet.
+///
+final class BottomSheetListSelectorViewController<DataSource: BottomSheetListSelectorDataSource, Model, Cell>:
+UIViewController, UITableViewDataSource, UITableViewDelegate where DataSource.Model == Model, DataSource.Cell == Cell {
+    private let viewProperties: BottomSheetListSelectorViewProperties
+    private var dataSource: DataSource
+    private let onDismiss: (_ selected: Model?) -> Void
+
+    private let rowType = Cell.self
+
+    private let estimatedSectionHeight = CGFloat(44)
+
+    @IBOutlet private(set) weak var tableView: UITableView!
+
+    init(viewProperties: BottomSheetListSelectorViewProperties,
+         dataSource: DataSource,
+         onDismiss: @escaping (_ selected: Model?) -> Void) {
+        self.viewProperties = viewProperties
+        self.dataSource = dataSource
+        self.onDismiss = onDismiss
+        super.init(nibName: "BottomSheetListSelectorViewController", bundle: nil)
+    }
+
+    required init?(coder: NSCoder) {
+        fatalError("init(coder:) has not been implemented")
+    }
+
+    override func viewDidLoad() {
+        super.viewDidLoad()
+
+        configureMainView()
+        configureTableView()
+    }
+
+    override func viewWillDisappear(_ animated: Bool) {
+        onDismiss(dataSource.selected)
+        super.viewWillDisappear(animated)
+    }
```

Comment:
This is a weird spot to call `onDismiss`. I wonder if `DrawerPresentationController` should provide this?

Reply:
right now bottom sheet dismiss is performed here on drag or tap outside of the sheet:

https://github.com/wordpress-mobile/WordPress-iOS/blob/3e51da289537324ce0cc27bcedf67db15d215091/WordPress/Classes/ViewRelated/System/Action%20Sheet/DrawerPresentationController.swift#L205-L207

I took a look at how `DrawerPresentationController` is used and we could pass a callback from `BottomSheetViewController` to `DrawerPresentationController`, but it also makes it hard to pass any other data from the content view controller (like passing the selected row from `BottomSheetListSelectorViewController`) because they are separate view controllers. do you think having `onDismiss` here is prone to errors?

Reply:
> I took a look at how DrawerPresentationController is used and we could pass a callback from BottomSheetViewController to DrawerPresentationController,

Looking at this bit of code:

https://github.com/woocommerce/woocommerce-ios/blob/55a1972194d311857571fdc7c6d9fa6874ba2be9/WooCommerce/Classes/ViewRelated/Products/ProductsViewController.swift#L442-L455

I think what feels a bit unnatural for me is that the `BottomSheetListSelectorViewController` emits the dismissal event (`onDismiss` block). However, `BottomSheetListSelectorViewController` is just a **child** of `BottomSheetViewController`. I feel like `BottomSheetViewController` should be telling us when it is dismissed or not. I expect that component to know when it's truly dismissed. And I would trust it more rather than the `childViewController`. Something like: 

```swift
        bottomSheet.onDismiss = { listSelectorViewController in
            let selected = listSelectorViewController.selected
        }
```

But for this to work well is that `BottomSheetViewController` should probably accept a type:

```swift
// is this how generics work? 😄 
class BottomSheetViewController<Child: UIViewController>: UIViewController {
```

Of course, the end result is still awkward because the dismiss is **still** in the `BottomSheetViewController`. We might still end up having a wrapper:

```swift
let bottomSheetListSelectorPresenter = BottomSheetListSelectorPresenter(command: command, onDismiss: { selected in
    // do something with `selected`
})
bottomSheetListSelectorPresenter.present(from: self, sourceView: sender)
```

With this, we've encapsulated all the wiring of the `BottomSheetViewController` in a single class (`*Presenter`) and there's less stuff to worry about inside `ProductsViewController`.

This requires quite a lot of changes so please just consider this as a thought exercise. What we have here is fine. 😄

Reply:
yea it does feel odd to handle the dismiss action from the child view controller, and `viewWillDisappear` is just a guess of the dismiss action and it's not impossible if some other events trigger it unexpectedly. I like the idea of having a wrapper like `BottomSheetListSelectorPresenter` in your example, and adding a completion block to the `dismiss` call in `DrawerPresentationController` for `BottomSheetListSelectorPresenter` to handle it. I added a subtask to https://github.com/woocommerce/woocommerce-ios/issues/2054 for improvements!